lib/fetch/index.js

```diff
@@ -126,8 +127,49 @@ class Fetch extends EE {
   }
 }
 
+// hoisting subscribers variables
+let subscribers
+
+// publishIfSubscribed will check if there are subscribers to the channel
+// add set additional context before publishing the event
+function publishIfSubscribed (name, context, type, load) {
+  if (subscribers[name]) {
+    if (type) context[type] = load
+    channels[name].publish(context)
+    if (type) context[type] = null
+  }
+}
+
+// handlePromise will handle both fulfilled/rejected promises from p
+// for diagnostics channel it will then check for subscribers and
+// publish events accordingly
+function handlePromise (p, context) {
+  return p.promise.then(
+    (success) => {
+      publishIfSubscribed('asyncStart', context)
+      publishIfSubscribed('asyncEnd', context)
+      return success
+    },
+    (failure) => {
+      publishIfSubscribed('asyncStart', context, 'error', failure)
+      publishIfSubscribed('asyncEnd', context, 'error', failure)
+      return Promise.reject(failure)
```

Comment:
can we somehow avoid to repackage the failure in a Promise.reject?

Reply:
This is to make sure that the failure is passed along do you have a suggestion how to do this differently?